neo/Consensus/ConsensusContext.cs

```diff
@@ -36,12 +36,18 @@ internal class ConsensusContext : IConsensusContext
         public Snapshot Snapshot { get; private set; }
         private KeyPair keyPair;
         private readonly Wallet wallet;
+        public Store Store { get; }
+        public bool IsRecovering { get; set; }
+        public HashSet<UInt256> KnownHashes { get; }
```

Comment:
`IsRecovering` and `KnownHashes` are never used in `ConsensusContext`, why move them into `ConsensusContext`? I think it's more reasonable to keep them in `ConsensusService`.

Reply:
I'm trying to reason on the precise importance of each class, and for me, ConsensusContext should hold the state of the system, while ConsensusService should manage the logic and may also hold important local variables.
I agree now that `KnownHashes` should stay on ConsensusService, because it's more like an auxiliar variable than actual state related.
In the case of `IsRecovering`, the original plan was to add that to the Helper class, so all important state functions would be there together. Do you think this solution is good Erik?

Reply:
`isRecovery` should stay on `ConsensusService`. Because it is just a flag of showing the begining and ending of `OnRecoveryMessageReceived()`.

Reply:
I see that it's a flag variable, but do we agree IsRecovery() is a state of the system? If the Helper manages to aggregate all states that's important, because we can easily create tests where the ConsensusService starts on Recovery, and if it becomes an internal variable of ConsensusService it may be harder to simulate such state.
But I may be wrong, I'll think more over this tomorrow, perhaps this is just an internal variable.

Reply:
https://github.com/neo-project/neo/blob/master/neo/Consensus/ConsensusService.cs#L309-312

It's atomic and temporary.

Reply:
I see now, you're correct as usual, this variable can only be `true` inside this block. However, side-effects like this and global variables are meant for great minds only, and my limited understanding usually requires variable passing by parameters, so I can actually know how this state is passed through. For example, I still don't get how `isRecovering` can be true on InitializeConsensus. All execution flows I followed do not allow that... can you or @jsolman clarify this please? Sorry for not following these dbft details earlier, but this is important for clear technical specification.

Reply:
`IsRecovering` is just used temporarily while a recovery message is being processed. Specifically, the timer needs to be set differently when recovering the primary from its own previously sent prepare request, since in that case it acts more like a backup in terms of its timeout.

Reply:
This thing to act like a `backup` is like a workaround in terms of flags, @igormcoelho , but, basically, Primary process its own preparereq without some verification.